source/shared/cpp/ObjectModel/MarkDownBlockParser.cpp

```diff
@@ -0,0 +1,417 @@
+}
+
+void EmphasisParser::CaptureCurrentCollectedStringAsRegularToken(std::string&currentToken) 
+{
+    if (currentToken.empty())
+        return;
+    std::shared_ptr<MarkDownStringHtmlGenerator> codeGen = 
+        std::make_shared<MarkDownStringHtmlGenerator>(currentToken);
+    std::shared_ptr<MarkDownHtmlGenerator> token = std::static_pointer_cast<MarkDownHtmlGenerator>(codeGen);
+    m_parsedResult.Append(token);
+
+    currentToken.clear();
+}
+
+void EmphasisParser::CaptureCurrentCollectedStringAsRegularToken() 
+{
+    if (m_current_token.empty())
+        return;
+    std::shared_ptr<MarkDownStringHtmlGenerator> codeGen = 
+        std::make_shared<MarkDownStringHtmlGenerator>(m_current_token);
+
+    std::shared_ptr<MarkDownHtmlGenerator> token = std::static_pointer_cast<MarkDownHtmlGenerator>(codeGen);
+    m_parsedResult.Append(token);
+
+    m_current_token.clear();
+}
+
+void EmphasisParser::UpdateCurrentEmphasisRunState(DelimiterType emphasisType)
+{
+    if (m_lookBehind != WhiteSpace)
+    {
+        m_checkLookAhead = (m_lookBehind == Puntuation);
+        m_checkIntraWord = (m_lookBehind == Alphanumeric && emphasisType == Underscore);
+    }
+    ++m_delimiterCnts;
+    m_currentDelimiterType = emphasisType;
+}
+
+bool EmphasisParser::IsRightEmphasisDelimiter(int ch)
+{
+    if (isspace(ch) && 
+       (m_lookBehind != WhiteSpace) && 
+       (m_checkLookAhead || m_checkIntraWord || m_currentDelimiterType == Asterisk))
+    {        
+        return true;
+    }
+
+    if (isalnum(ch) && 
+        m_lookBehind != WhiteSpace && 
+        m_lookBehind != Init)
+    {
+        if (!m_checkLookAhead && !m_checkIntraWord)
+        {
+            return true;
+        }
+
+        if (m_checkLookAhead || m_checkIntraWord)
+        {
+            return false;
+        }
+    }
+
+    if (ispunct(ch) && m_lookBehind != WhiteSpace)
+    {
+        return true;
+    }
+
+    return false;;
+}
+
+bool EmphasisParser::TryCapturingRightEmphasisToken(int ch, std::string &currentToken)
+{
+    if (IsRightEmphasisDelimiter(ch))
+    {
+        std::shared_ptr<MarkDownRightEmphasisHtmlGenerator> codeGen = nullptr;
+        if (IsLeftEmphasisDelimiter(ch))
+        {
+            codeGen = 
+                std::make_shared<MarkDownLeftAndRightEmphasisHtmlGenerator>(currentToken, m_delimiterCnts, 
+                        m_currentDelimiterType);
+        }
+        else
+        {
+            codeGen = 
+                std::make_shared<MarkDownRightEmphasisHtmlGenerator>(currentToken, m_delimiterCnts, 
+                        m_currentDelimiterType);
+        }
+
+        std::shared_ptr<MarkDownEmphasisHtmlGenerator> emphasisToken = std::static_pointer_cast<MarkDownEmphasisHtmlGenerator>(codeGen);
+        m_parsedResult.Append(emphasisToken);
+
+        std::shared_ptr<MarkDownHtmlGenerator> htmlToken = std::static_pointer_cast<MarkDownHtmlGenerator>(codeGen);
+        m_parsedResult.Append(htmlToken);
```

Comment:
See my comment in MarkDownParsedResult.cpp.  Once you make that change you should only need one variable here and you shouldn't need any of these casts.

Reply:
Same in TryCapturingLeftEmphasisToken

---
In reply to: [157612265](https://github.com/Microsoft/AdaptiveCards/pull/1045#discussion_r157612265) [](ancestors = 157612265)